src/platform/windows/display_wgc.cpp

```diff
@@ -0,0 +1,290 @@
+        BOOST_LOG(error) << "Failed to query DXGI interface from device [0x"sv << util::hex(status).to_string_view() << ']';
+        return -1;
+      }
+      if (FAILED(status = winrt::CreateDirect3D11DeviceFromDXGIDevice(*&dxgi, d3d_comhandle.put()))) {
+        BOOST_LOG(error) << "Failed to query WinRT DirectX interface from device [0x"sv << util::hex(status).to_string_view() << ']';
+        return -1;
+      }
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: "sv << e.code();
+      return -1;
+    }
+
+    DXGI_OUTPUT_DESC output_desc;
+    uwp_device = d3d_comhandle.as<winrt::IDirect3DDevice>();
+    display->output->GetDesc(&output_desc);
+
+    auto monitor_factory = winrt::get_activation_factory<winrt::GraphicsCaptureItem, IGraphicsCaptureItemInterop>();
+    if (monitor_factory == nullptr ||
+        FAILED(status = monitor_factory->CreateForMonitor(output_desc.Monitor, winrt::guid_of<winrt::IGraphicsCaptureItem>(), winrt::put_abi(item)))) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: [0x"sv << util::hex(status).to_string_view() << ']';
+      return -1;
+    }
+
+    if (config.dynamicRange)
+      display->capture_format = DXGI_FORMAT_R16G16B16A16_FLOAT;
+    else
+      display->capture_format = DXGI_FORMAT_B8G8R8A8_UNORM;
+
+    try {
+      frame_pool = winrt::Direct3D11CaptureFramePool::CreateFreeThreaded(uwp_device, static_cast<winrt::Windows::Graphics::DirectX::DirectXPixelFormat>(display->capture_format), 2, item.Size());
+      capture_session = frame_pool.CreateCaptureSession(item);
+      frame_pool.FrameArrived({ this, &wgc_capture_t::on_frame_arrived });
+      capture_session.IsBorderRequired(false);
+      capture_session.StartCapture();
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(error) << "Screen capture is not supported on this device for this release of Windows: "sv << e.code();
+      return -1;
+    }
+    return 0;
+  }
+
+  void
+  wgc_capture_t::on_frame_arrived(winrt::Direct3D11CaptureFramePool const &sender, winrt::IInspectable const &) {
+    // this PRODUCER runs in a separate thread spawned by the frame pool. to
+    // retain parity with the original interface, the frame will be consumed by
+    // the capture thread, not this one.
+
+    winrt::Windows::Graphics::Capture::Direct3D11CaptureFrame frame { nullptr };
+    try {
+      frame = sender.TryGetNextFrame();
+    }
+    catch (winrt::hresult_error &e) {
+      BOOST_LOG(warning) << "Failed to capture frame: "sv << e.code();
+      return;
+    }
+    if (frame != nullptr) {
+      AcquireSRWLockExclusive(&frame_lock);
+      if (produced_frame) {
+        produced_frame.Close();
+      }
+      produced_frame = frame;
+      WakeConditionVariable(&frame_present_cv);
+      ReleaseSRWLockExclusive(&frame_lock);

```

Comment:
Microsoft advises that waking outside the lock is generally the optimal ordering.

> You can wake other threads using [WakeConditionVariable](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-wakeconditionvariable) or [WakeAllConditionVariable](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-wakeallconditionvariable) either inside or outside the lock associated with the condition variable. It is usually better to release the lock before waking other threads to reduce the number of context switches.